### PR TITLE
Homework Week 5 - mandrulli

### PR DESCRIPTION
**Pull request by mandrulliBF**

### CODE REVIEW

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

Overall, your solution is well-structured and logically follows the given homework prompt. You have successfully implemented the state change tracking for `nba_players` as required. However, there are a few areas where your solution could be improved.

1) **Covering all aspects of the homework prompt:** Your solution covers most aspects of the prompt, but it seems to be missing the creation of the `players_state_tracking` table. The example solution starts by creating this table, which is then used to store the state change tracking data. You might want to consider adding this step to your solution.

2) **Readability of the code:** Your code is generally well-structured and easy to follow. You have used meaningful variable names and included comments to explain your thought process, which is great. However, the readability could be improved by adding more comments to explain the logic behind your `CASE` statements, especially the one used to determine the `player_state`.

3) **Robustness and error handling:** Your solution seems to handle the given data well, but it's not clear how it would handle edge cases or invalid inputs. For example, how would your code handle a player who was active in the previous season but is not present in the current season data? It would be beneficial to add some error handling or checks to ensure the data is as expected.

4) **Modularity and reusability of the code:** Your code is well-structured and the use of CTEs (Common Table Expressions) makes it modular and easy to understand. However, the `CASE` statement used to determine the `player_state` is quite complex and could be simplified or broken down into smaller parts to improve reusability.

Here are some specific suggestions for improvement:

- Consider creating the `players_state_tracking` table at the beginning of your solution, as shown in the example solution.
- Add more comments to explain the logic behind your `CASE` statements, especially the one used to determine the `player_state`.
- Consider adding some error handling or checks to ensure the data is as expected.
- Try to simplify or break down the `CASE` statement used to determine the `player_state` to improve reusability.

Keep up the good work! Your solution is on the right track, and with a few improvements, it could be even better.

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

Overall, your solution is well-structured and logically follows the assignment prompt. You've done a good job of using SQL to track the state changes of NBA players. However, there are a few areas where your solution could be improved.

1) **Completeness of the solution:** Your solution does not cover all the aspects of the homework prompt. The prompt asks for a state change-tracking field that takes on the values `New`, `Retired`, `Continued Playing`, `Returned from Retirement`, and `Stayed Retired`. Your solution does not seem to handle the case where a player stays out of the league, which should be marked as `Stayed Retired`. 

2) **Readability of the code:** Your code is generally well-structured and easy to read. You've done a good job of using meaningful variable names and including comments to explain your thought process. However, the logic in your CASE statement could be simplified to improve readability. For example, you could use a single CASE statement to determine the `player_state` based on the conditions outlined in the prompt, similar to the example solution.

3) **Robustness and error handling:** Your solution does not seem to handle potential errors or edge cases. For example, what happens if the `player_name` is NULL or if the `season` is not a valid year? It would be beneficial to add some error handling to your code to ensure it can handle these situations gracefully.

4) **Modularity and reusability:** Your code is structured in a way that makes it easy to understand and reuse specific components in other contexts. The use of CTEs (Common Table Expressions) like `last_season` and `current_season` makes your code modular and easy to modify if needed.

To improve your solution, I would suggest the following:

- Add a condition to handle the case where a player stays out of the league, which should be marked as `Stayed Retired`.
- Simplify the logic in your CASE statement to improve readability.
- Add error handling to your code to ensure it can handle potential errors or edge cases.

Keep up the good work! Your understanding of SQL and ability to apply it to solve complex problems is evident in your solution. With a few improvements, your solution will be even better.

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_2.sql`

Overall, your solution is well-structured and logically follows the given homework prompt. You have correctly used the `GROUPING SETS` clause to perform aggregations along the required dimensions. However, there are a few areas where your solution could be improved.

1) **Covering all aspects of the prompt:** Your solution does not fully cover the prompt. The prompt asks for aggregations along the dimensions of player and team, player and season, and team. While you have correctly implemented these groupings, the prompt also asks for the creation of a new table `game_details_dashboard`, which your solution does not include. You can add this by wrapping your query in a `CREATE TABLE game_details_dashboard AS` statement.

2) **Readability:** Your code is generally well-formatted and easy to read. However, you could improve readability by adding comments to explain your thought process and the purpose of each part of your code. This will make it easier for others (and your future self) to understand what your code is doing.

3) **Robustness and error handling:** Your code does not handle potential errors or edge cases. For example, you do not handle the case where `games.home_team_id`, `games.visitor_team_id`, `home_team_wins`, or `dedup.team_id` could be `NULL`. This could lead to incorrect results or errors. You should add checks or conditions to handle these potential issues.

4) **Modularity and reusability:** Your code is modular and the use of a CTE (Common Table Expression) makes it easy to understand and reuse specific components. However, you could improve the reusability of your code by defining the CTE as a view, which can be reused in other queries.

Here's how you could modify your code to address these points:

```sql
CREATE TABLE game_details_dashboard AS
WITH combined AS (
    SELECT
      games.season,
      dedup.player_name AS player,
      dedup.team_abbreviation AS team,
      dedup.pts AS pts,
      CASE
        WHEN games.home_team_id = dedup.team_id AND home_team_wins = 1 THEN games.game_id
        WHEN games.visitor_team_id = dedup.team_id AND home_team_wins = 0 THEN games.game_id
        WHEN home_team_wins IS NULL THEN NULL
        ELSE NULL
      END AS match_won
    FROM
      bootcamp.nba_game_details_dedup AS dedup
      JOIN bootcamp.nba_games AS games ON games.game_id = dedup.game_id
    WHERE games.home_team_id IS NOT NULL AND games.visitor_team_id IS NOT NULL AND home_team_wins IS NOT NULL AND dedup.team_id IS NOT NULL
)
SELECT
  COALESCE(player, 'overall') AS player,
  COALESCE(CAST(season AS VARCHAR), 'overall') AS season,
  COALESCE(team, 'overall') AS team,
  SUM(pts) AS total_points,
  COUNT(DISTINCT match_won) AS total_wins
FROM
  combined
GROUP BY
  GROUPING SETS ((player, team), (player, season), (team)) -- slices that aggregate along the following combinations
```

Keep up the good work, and remember to always consider edge cases and potential errors in your code.

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_2.sql`

Overall, your solution is well-structured and logically follows the given homework prompt. You have correctly used the `GROUPING SETS` clause to perform aggregations along the required dimensions. However, there are a few areas where your solution could be improved.

1) **Missing Aggregation Level:** The assignment prompt asks for an `aggregation_level` column in the output, which your solution does not provide. This column is used to indicate the level of aggregation for each row in the output. You can add this column to your output by using the `GROUPING` function, similar to the example solution.

2) **Win Calculation:** Your calculation for `match_won` seems to be incorrect. You are assigning the `game_id` to `match_won` when a team wins, but the assignment prompt asks for a binary value indicating whether the team won or not. You should adjust your `CASE` statement to assign 1 when a team wins and 0 otherwise.

3) **Error Handling:** Your solution does not handle potential errors or edge cases. For example, you are assuming that `home_team_id`, `visitor_team_id`, `home_team_wins`, and `team_id` are never `NULL`. It would be better to handle these potential `NULL` values in your code.

4) **Code Readability:** Your code is generally well-formatted and easy to read. However, you could improve the readability of your code by adding comments to explain your thought process and the purpose of each section of your code.

5) **Modularity and Reusability:** Your code is modular and the `WITH` clause makes it easy to understand and reuse specific components in other contexts. However, you could improve the reusability of your code by defining the `GROUPING SETS` as a variable at the beginning of your code, which would make it easier to modify the dimensions of aggregation in the future.

Here is a revised version of your solution incorporating these suggestions:

```sql
CREATE TABLE game_details_dashboard AS
WITH combined AS (
    SELECT
      games.season,
      dedup.player_name AS player,
      dedup.team_abbreviation AS team,
      dedup.pts AS pts,
      CASE
        WHEN games.home_team_id = dedup.team_id AND home_team_wins = 1 THEN 1
        WHEN games.visitor_team_id = dedup.team_id AND home_team_wins = 0 THEN 1
        ELSE 0
      END AS match_won
    FROM
      bootcamp.nba_game_details_dedup AS dedup
    JOIN bootcamp.nba_games AS games ON games.game_id = dedup.game_id
    WHERE games.home_team_id IS NOT NULL AND games.visitor_team_id IS NOT NULL AND home_team_wins IS NOT NULL AND dedup.team_id IS NOT NULL
)
SELECT
  CASE 
    WHEN GROUPING(player, team) = 0 THEN 'player_team'
    WHEN GROUPING(player, season) = 0 THEN 'player_season'
    WHEN GROUPING(team) = 0 THEN 'team'
    ELSE 'overall'
  END as aggregation_level,
  COALESCE(player, 'overall') AS player,
  COALESCE(CAST(season AS VARCHAR), 'overall') AS season,
  COALESCE(team, 'overall') AS team,
  SUM(pts) AS total_points,
  SUM(match_won) AS total_wins
FROM
  combined
GROUP BY
  GROUPING SETS ((player, team), (player, season), (team)) -- slices that aggregate along the following combinations
```

Keep up the good work!

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_2.sql`

Overall, your solution is well-structured and logically follows the given homework prompt. You have successfully used `GROUPING SETS` to perform aggregations of the `game_details` data along the required dimensions. However, there are a few areas where your solution could be improved.

1) **Readability and Comments**: Your code is generally well-formatted and easy to read. However, it lacks comments explaining your thought process. Adding comments to your code can make it easier for others (and your future self) to understand what you're trying to achieve. For example, you could add comments explaining the purpose of the `combined` CTE and the logic behind your `CASE` statement in the `SELECT` clause.

2) **Robustness and Error Handling**: Your code includes a `WHERE` clause in the `combined` CTE to filter out rows where certain fields are `NULL`. This is a good practice as it ensures that your code won't break due to missing data. However, it would be even better if you could handle these cases in a more robust way. For example, you could use the `COALESCE` function to replace `NULL` values with a default value.

3) **Modularity and Reusability**: Your code is well-structured and the use of a CTE makes it modular and easy to understand. However, the `GROUP BY GROUPING SETS` clause could be made more reusable by defining the grouping sets as a variable at the beginning of your code. This would make it easier to modify the grouping sets if needed.

4) **Aggregation of Wins**: In your code, you're using `COUNT(DISTINCT match_won)` to calculate the total number of wins. However, this might not give the correct result if a team wins more than one match with the same `game_id`. Instead, you could use `SUM(match_won)` as in the example solution, where `match_won` is a binary field indicating whether the team won or not.

5) **Variable Names**: Your variable names are generally meaningful and descriptive. However, the name `match_won` might be a bit misleading as it suggests that it's a binary field indicating whether the match was won or not, while in fact it contains the `game_id` of the won matches. A more descriptive name could be `won_game_id`.

In conclusion, your solution is quite good, but there's room for improvement in terms of readability, robustness, reusability, and accuracy of the results. Keep up the good work!

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_2.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_3.sql`

Overall, your solution is quite good and it logically follows the given homework prompt. You have successfully written a query to find out which player scored the most points playing for a single team. However, there are a few areas where you could improve your code.

1) **Variable Names:** You have used `player` and `team` as variable names, while the prompt and example solution use `player_name` and `team_name`. It's important to use meaningful and consistent variable names to improve the readability of your code. I would suggest changing `player` to `player_name` and `team` to `team_name` to match the example solution and the prompt.

2) **Comments:** You have included a comment at the beginning of your code, which is good practice. However, it would be beneficial to include more comments throughout your code to explain your thought process and make your code easier to understand for others (and for yourself when you revisit it in the future).

3) **Robustness and Error Handling:** Your code does not handle potential errors or edge cases. For example, what if the `total_points` field is not a number or is negative? It would be good to add some error handling or checks to ensure the data is as expected.

4) **Modularity and Reusability:** Your code is quite modular and could be easily adapted for similar queries. However, you could improve the reusability of your code by defining some of the strings (like 'player_team') as variables at the top of your code. This would make it easier to modify your code for similar queries in the future.

Here's how your improved code might look:

```sql
-- Define variables
DEFINE AGGREGATION_LEVEL = 'player_team';

-- Which player scored the most points playing for a single team?
SELECT player_name, team_name, total_points
FROM game_details_dashboard
WHERE aggregation_level = AGGREGATION_LEVEL 
AND total_points IS NOT NULL
AND total_points >= 0 -- Ensure points are non-negative
ORDER BY total_points DESC 
LIMIT 1
```

Keep up the good work! With a few minor improvements, your code will be even better.

### Feedback for `query_4.sql`

Overall, your solution is quite good and it logically follows the given homework prompt. You have correctly used the `GROUPING SETS` aggregation to find the player who scored the most points in one season. However, there are a few areas where your solution could be improved.

1) **Coverage of the Homework Prompt:** The homework prompt asks for the player who scored the most points in one season, but your query also includes the team. While this is not incorrect, it is not necessary for the prompt. You could simplify your query by removing the `team` from the `SELECT` clause.

2) **Readability:** Your code is generally well-formatted and easy to read. However, you could improve the readability of your code by adding comments to explain your thought process. For example, you could explain why you are filtering for `total_points IS NOT NULL` and `aggregation_level = 'player_season'`.

3) **Robustness and Error Handling:** Your code does a good job of handling potential errors by checking that `total_points IS NOT NULL`. This will prevent any errors that could occur from trying to order by null values. However, you could improve your error handling by also checking that `player` is not null. This would prevent any errors that could occur if there were rows in the table where the player was null.

4) **Modularity and Reusability:** Your code is quite modular and could easily be reused in other contexts. However, you could improve the reusability of your code by parameterizing the `aggregation_level`. This would allow you to easily change the level of aggregation without having to modify the code.

Here's how you could implement these suggestions:

```sql
-- Which player scored the most points in one season?

-- We filter for rows where total_points and player are not null to prevent errors
-- We also filter for rows where aggregation_level is 'player_season' to get the total points for each player in each season
-- We then order by total_points in descending order and limit the results to 1 to get the player with the most points

SELECT player, total_points
FROM game_details_dashboard
WHERE aggregation_level = 'player_season'
AND total_points IS NOT NULL
AND player IS NOT NULL
ORDER BY total_points DESC 
LIMIT 1
```

Keep up the good work! With a few minor improvements, your code will be even better.

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_2.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_3.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_4.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_5.sql`

Overall, your solution is well-structured and logically follows the given homework prompt. However, there are a few areas where it could be improved. Here's my feedback:

1) **Coverage of the Homework Prompt:** Your code does not fully cover the homework prompt. The prompt asks for the team that has won the most games, but your query is filtering on `aggregation_level = 'team'` instead of `aggregation_level = 'player_name__team_name'` as specified in the example solution. This might lead to different results depending on how the data is structured in the `game_details_dashboard` table. Please adjust your query to match the prompt.

2) **Logical Flow:** Your solution logically follows the prompt. You correctly identified the need to order the results by `total_wins` in descending order and limit the output to the top result. This is the correct approach to identify the team with the most wins.

3) **Readability:** Your code is generally readable and well-formatted. You've used meaningful variable names and included a comment at the beginning of your query, which is good practice. However, it would be helpful to include additional comments explaining why you're filtering on `total_wins IS NOT NULL`. This would make it easier for others to understand your thought process.

4) **Robustness and Error Handling:** Your code includes a check for `total_wins IS NOT NULL`, which is a good practice for handling potential NULL values in the data. However, it's not clear why this check is necessary based on the provided prompt. If there's a specific reason for this check, please include it in your comments.

5) **Modularity and Reusability:** Your code is structured in a way that makes it easy to understand and reuse. The query could be easily modified to answer similar questions about the data, such as "Which team has lost the most games?" or "Which team has the most ties?".

In summary, your solution is on the right track, but needs a few adjustments to fully meet the requirements of the prompt. Keep up the good work!

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_2.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_3.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_4.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_5.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_7.sql`

Overall, your solution is well-structured and logically follows the homework prompt. However, there are a few areas where it could be improved to better meet the requirements of the prompt and to enhance readability and robustness.

1) **Coverage of the Homework Prompt:** Your solution does not fully cover the prompt. The prompt asks for the number of games in a row where LeBron James scored over 10 points. Your solution calculates the total number of games where LeBron James scored over 10 points in consecutive games, but it does not identify the longest streak of such games. 

To fix this, you could modify your approach to track the streaks of games where LeBron James scored over 10 points. One way to do this would be to create a new group every time `plus_10_pts` changes from `true` to `false` or vice versa. Then, you could count the number of games in each group and return the maximum count.

2) **Readability:** Your code is generally well-formatted and easy to read. You've used meaningful variable names and included comments to explain your thought process, which is great. However, the comment on the `LAG` function could be more clear. Instead of saying "Get player scoring > 10 pts in the last game", you could say "Check if the player scored > 10 pts in the previous game".

3) **Robustness and Error Handling:** Your code does not handle the case where LeBron James did not play in a game. If LeBron did not play, he would not have scored any points, which would break his streak. You could handle this by adding a condition to your `WHERE` clause to only include games where LeBron James played.

4) **Modularity and Reusability:** Your code is modular and could be easily adapted to answer similar questions about other players. However, the `LAG` function and the `IF` statement in the `SUM` function are specific to the current question and may not be applicable in other contexts. You could increase the reusability of your code by moving these parts into separate subqueries or functions.

Here's an example of how you could modify your code to address these points:

```sql
WITH nba_games_data AS (
    SELECT 
        dedup.game_id, 
        dedup.player_name,
        dedup.pts > 10 AS plus_10_pts,
        games.game_date_est AS game_date
    FROM bootcamp.nba_game_details_dedup AS dedup
    JOIN bootcamp.nba_games AS games ON games.game_id = dedup.game_id
    WHERE dedup.player_name = 'LeBron James' AND dedup.min > 0 -- Only include games where LeBron James played
),
lagged AS (
    SELECT 
        *,
        LAG(plus_10_pts, 1) OVER (PARTITION BY player_name ORDER BY game_date ASC) AS last_game_plus_10_pts -- Check if the player scored > 10 pts in the previous game
    FROM nba_games_data
),
streaks AS (
    SELECT 
        *,
        SUM(IF(plus_10_pts AND NOT last_game_plus_10_pts, 1, 0)) OVER (PARTITION BY player_name ORDER BY game_date ASC) AS streak_group -- Create a new group every time plus_10_pts changes from true to false or vice versa
    FROM lagged
)
SELECT
    player_name,
    MAX(COUNT(*)) OVER (PARTITION BY player_name, streak_group) AS max_streak_plus_10_pts -- Count the number of games in each streak and return the maximum count
FROM streaks
WHERE plus_10_pts
GROUP BY player_name, streak_group
ORDER BY max_streak_plus_10_pts DESC
LIMIT 1
```
This code should return the longest streak of games where LeBron James scored over 10 points.

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_2.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_3.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_4.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_5.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_6.sql`

Your solution to the problem is quite good, and you've done a great job using window functions to calculate the rolling sum of wins over a 90-game stretch. However, there are a few areas where your solution could be improved.

1) **Completeness of the solution:** Your solution does not fully address the prompt. The prompt asks for the most games a single team has won in a given 90-game stretch, but your solution only provides the team that has won the most games. To provide a complete solution, you should also include the specific 90-game stretch during which the team won the most games. You can do this by including the start and end dates of the 90-game stretch in your final SELECT statement.

2) **Logical flow:** Your solution logically follows the prompt, but there is a potential issue in your CASE statement. You are assigning the game_id to the match_won column when a team wins, and then summing this column to get the total number of wins. This could lead to incorrect results if the game_id values are not 1 for each win. Instead, you should assign 1 when a team wins and 0 when it loses, similar to the example solution.

3) **Readability:** Your code is generally well-structured and easy to read. However, you could improve the readability by adding more comments to explain your thought process and the purpose of each part of your code. For example, you could explain why you are using a CASE statement and what the SUM window function is doing.

4) **Robustness and error handling:** Your code does not handle potential errors or edge cases. For example, if the home_team_wins column is NULL, your CASE statement assigns NULL to the match_won column. This could lead to incorrect results when you sum this column. You should handle this case by assigning 0 to the match_won column when home_team_wins is NULL.

5) **Modularity and reusability:** Your code is modular and the use of CTEs (Common Table Expressions) makes it easy to understand and reuse specific components. However, you could improve the reusability of your code by defining a function to calculate the rolling sum of wins, which could be reused in other queries.

Overall, you've done a good job, but there are a few areas where you could improve your solution. Keep practicing and keep up the good work!

### Feedback for `query_7.sql`

No feedback generated as no changes were detected in the current push.

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_2.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_3.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_4.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_5.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_6.sql`

The student's solution appears to cover all aspects of the homework prompt. The student has correctly used window functions to calculate the maximum number of games won by a team in a 90-game stretch. The logic of the student's solution aligns with the given homework prompt, and the overall structure of the code is coherent and easy to follow.

The readability of the code is good. The student has used meaningful variable names that clearly indicate what each variable represents. The use of comments to explain the purpose of each section of the code is commendable and enhances the readability of the code. The formatting of the code is also well done, with proper indentation and spacing.

The robustness and error handling of the code seem to be adequate. The student has used the `DISTINCT` keyword to avoid duplicate records, which is a good practice. However, the student could improve the robustness of the code by adding more error handling, such as checking for null values or invalid inputs.

The modularity and reusability of the code are good. The student has used Common Table Expressions (CTEs) to break down the problem into smaller, more manageable parts. This makes the code easier to understand and reuse in other contexts.

One suggestion for improvement would be to handle the case when `home_team_wins` is `NULL` in the `nba_games_data` CTE. Currently, the student's code assigns `NULL` to `match_won` when `home_team_wins` is `NULL`, which could potentially lead to incorrect results in the `ninety_days_wins_streak` calculation. A better approach might be to exclude these records from the calculation altogether.

Overall, the student's solution is well-structured, readable, and logically sound. With a few minor improvements, the code could be even more robust and error-proof.

### Feedback for `query_7.sql`

No feedback generated as no changes were detected in the current push.

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_2.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_3.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_4.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_5.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_6.sql`

Overall, your solution is well-structured and logically follows the given homework prompt. You have successfully used window functions to calculate the most games a single team has won in a 90-game stretch. However, there are a few areas where your solution could be improved.

1) **Coverage of the homework prompt:** Your solution covers most aspects of the homework prompt, but it seems to be missing the handling of the case where the team is the visitor and wins the game. In your `CASE` statement, you only handle the case where the team is the home team and wins, and the case where the team is the visitor and loses. You should add another condition to handle the case where the team is the visitor and wins. Here is how you could modify your `CASE` statement:

```sql
CASE
    WHEN games.home_team_id = dedup.team_id AND home_team_wins = 1 THEN 1
    WHEN games.visitor_team_id = dedup.team_id AND home_team_wins = 0 THEN 1
    WHEN home_team_wins IS NULL THEN NULL
ELSE NULL
END AS match_won
```

2) **Readability:** Your code is generally well-formatted and easy to read. You have used meaningful variable names and included comments to explain your thought process, which is great. However, the comment on the line calculating the rolling sum could be improved for clarity. Instead of "Rolling sum over the last 90 game", consider "Rolling sum of wins over the last 90 games".

3) **Robustness and error handling:** Your code does not seem to handle potential errors or edge cases. For example, what happens if `home_team_wins` is not 1 or 0? What if `game_date` is NULL? Consider adding error handling or checks for these potential issues.

4) **Modularity and reusability:** Your code is well-structured and the use of CTEs (Common Table Expressions) enhances its modularity and reusability. Each CTE is responsible for a specific part of the logic, which makes your code easier to understand and modify.

Keep up the good work, and consider these suggestions for improvement in your future assignments.

### Feedback for `query_7.sql`

No feedback generated as no changes were detected in the current push.

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_2.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_3.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_4.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_5.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_6.sql`

Overall, your solution is well-structured and logically follows the given homework prompt. You have successfully used window functions to calculate the most games a single team has won in a 90-game stretch. However, there are a few areas where your solution could be improved.

1) **Coverage of the homework prompt:** Your solution covers most aspects of the homework prompt, but it seems to be missing the part where you need to calculate the most games a single team has won in a given 90-game stretch. Your current solution calculates the most games won in a 90-day stretch, which is not the same as a 90-game stretch. To fix this, you should adjust your window function to count the number of games won in the last 90 games, not the last 90 days.

2) **Readability of the code:** Your code is generally well-formatted and easy to read. You have used meaningful variable names and included comments to explain your thought process, which is great. However, the comment on the line calculating the rolling sum could be more clear. Instead of saying "Rolling sum over the last 90 game", it would be more accurate to say "Rolling sum of wins over the last 90 games".

3) **Robustness and error handling:** Your code does not seem to handle the case where `home_team_wins` is NULL. In your CASE statement, you have a condition `WHEN home_team_wins IS NULL THEN NULL`, but you do not handle this NULL value later in your code. This could potentially lead to errors or incorrect results. You should either handle this NULL value appropriately or filter out these records before performing your calculations.

4) **Modularity and reusability of the code:** Your code is well-structured and modular, with separate CTEs for different parts of the calculation. This makes it easy to understand and reuse specific components in other contexts. However, the `nba_games_data` CTE could be made more reusable by removing the specific calculation of `match_won` and moving it to a separate CTE or the main query. This would allow the `nba_games_data` CTE to be used for other calculations involving NBA games data.

In conclusion, your solution is on the right track, but needs a few adjustments to fully meet the requirements of the homework prompt and to improve its readability, robustness, and reusability. Keep up the good work!

### Feedback for `query_7.sql`

No feedback generated as no changes were detected in the current push.

File: submission/query_1.sql [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+combined AS (
+    SELECT
+        COALESCE(ls.player_name, cs.player_name) AS player_name,
+        COALESCE(ls.first_active_season, IF(cs.is_active, cs.current_season, NULL)) AS first_active_season, -- Get if this season was the first active
+        COALESCE(IF(cs.is_active, cs.current_season, NULL), ls.last_active_season) AS last_active_season, -- Get if this season was the last active 
+        CASE
+            WHEN ls.seasons_active IS NULL AND cs.is_active THEN ARRAY[cs.current_season]
+            WHEN ls.seasons_active IS NOT NULL AND (cs.is_active IS NULL OR NOT cs.is_active) THEN ls.seasons_active
+            ELSE ls.seasons_active || ARRAY[cs.current_season] 
+        END AS seasons_active, -- Array containing all the active seasons
```
Comment (HjersmanJ) [inline submission/query_1.sql:26]: Some of the comments are superfluous like this one. Would recommend omitting comments that only describe a simple portion of the code.

File: submission/query_2.sql [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+      CASE
+        WHEN games.home_team_id = dedup.team_id AND home_team_wins = 1 THEN games.game_id
+        WHEN games.visitor_team_id = dedup.team_id AND home_team_wins = 0 THEN games.game_id
+        WHEN home_team_wins IS NULL THEN NULL
+        ELSE NULL
+      END AS match_won
+    FROM
+      bootcamp.nba_game_details_dedup AS dedup
+    JOIN bootcamp.nba_games AS games ON games.game_id = dedup.game_id
+    WHERE games.home_team_id IS NOT NULL AND games.visitor_team_id IS NOT NULL AND home_team_wins IS NOT NULL AND dedup.team_id IS NOT NULL
```
Comment (HjersmanJ) [inline submission/query_2.sql:17]: Good job deduping.

File: submission/query_5.sql
Comment (HjersmanJ) [inline submission/query_5.sql:1]: Good job on the grouping set queries.

File: submission/query_6.sql [hunk 3]
```diff
@@ -0,0 +1,30 @@
+WITH nba_games_data AS (
+SELECT DISTINCT
+    dedup.game_id,
+    dedup.team_abbreviation AS team,
+    CASE
+        WHEN games.home_team_id = dedup.team_id AND home_team_wins = 1 THEN 1
+        WHEN games.visitor_team_id = dedup.team_id AND home_team_wins = 0 THEN 1
+        WHEN home_team_wins IS NULL THEN NULL
+    ELSE NULL
+    END AS match_won, -- The team won the play
```
Comment (HjersmanJ) [inline submission/query_6.sql:10]: This comment is fine.

**Comment by julie-scherer** [comment]
## Automated Feedback from ChatGPT

### Feedback for `query_1.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_2.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_3.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_4.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_5.sql`

No feedback generated as no changes were detected in the current push.

### Feedback for `query_6.sql`

The student's solution appears to cover all aspects of the homework prompt. The student has correctly used window functions to calculate the maximum number of games won by a team in a 90-game stretch. The logic of the student's solution aligns with the given homework prompt, and the overall structure of the code is coherent and easy to follow.

The readability of the student's code is good. The student has used meaningful variable names and included comments to explain their thought process. The formatting of the code is also well done, making it easy to read and understand.

The student's code seems to handle edge cases and potential errors well. The use of `DISTINCT` in the first `WITH` clause ensures that duplicate records are not considered. The `CASE` statement in the `nba_games_data` subquery handles different scenarios of a team winning a game. However, the student could improve error handling by adding more comments to explain the logic behind the `CASE` statement and what each condition represents.

The modularity and reusability of the student's code are commendable. The student has broken down the problem into two subproblems, each solved in a separate `WITH` clause. This makes the code easier to understand and modify. The `nba_games_data` subquery can be reused to analyze other aspects of the games, and the `streak_wins_over_90` subquery can be adapted to analyze winning streaks over different numbers of games.

Here are a few suggestions for improvement:

1) In the `nba_games_data` subquery, the `ELSE NULL` clause in the `CASE` statement is redundant and can be removed. If none of the conditions in the `CASE` statement are met, SQL will automatically return `NULL`.

2) In the `streak_wins_over_90` subquery, the student could add a `WHERE` clause to filter out records where `match_won` is `NULL` before calculating the rolling sum. This would make the code more robust by ensuring that only valid records are considered.

3) The student could add more comments to explain the logic behind the `CASE` statement in the `nba_games_data` subquery and what each condition represents. This would make the code easier to understand for other people reading it.

4) The student could consider adding a `HAVING` clause to the final `SELECT` statement to filter out teams that have not won any games in a 90-game stretch. This would make the results more meaningful and easier to interpret.

### Feedback for `query_7.sql`

No feedback generated as no changes were detected in the current push.

File: submission/query_6.sql
Comment (HjersmanJ) [inline submission/query_6.sql:1]: Close! You are 3 over the correct answer. I suspect it's because you didn't dedup the data.

File: submission/query_7.sql
Comment (HjersmanJ) [inline submission/query_7.sql:1]: Answer I get from this query is 1623. Lebron hasn't even played 1500 games. This is likely due to you not deduping the data. Correct answer is in the 60's.

**Review by HjersmanJ — APPROVED** [review]
Your code is clean and accurate for the most part. The major issue is not deduping the data for the later queries. Query 6 is hard to spot but query 7 should cause some surprises. It's okay to ask the chat or "colleagues" about this result even if you don't have domain knowledge (i.e. basketball). 

What would make this assignment complete is to include in the README how to use the grouping sets. Describe the aggregation level grains and what they are and their purpose. Otherwise, great job!
